File: sorting/java/MergeSort.java
/**
 * Merge Sort - Sorting Algorithm
 *
 * Time Complexity:
 * - Best case: O(n log n)
 * - Average case: O(n log n)
 * - Worst case: O(n log n)
 *
 * Space Complexity: O(n) - requires additional space for temporary arrays
 *
 * How it works:
 * Merge Sort is a divide and conquer sorting algorithm.
 * 1. Divide the array into two halves
 * 2. Recursively sort each half
 * 3. Merge the two sorted halves to produce the final sorted array
 */
import java.util.Arrays;

public class MergeSort {
    
    /**
     * Implementation of the Merge Sort algorithm for integers.
     * 
     * @param arr Array to be sorted
     * @param reverse If true, sorts in descending order; if false, in ascending order
     * @return Sorted array
     */
    public static int[] mergeSort(int[] arr, boolean reverse) {
        // Create a copy of the array to avoid modifying the original
        int[] result = Arrays.copyOf(arr, arr.length);
        
        // Start the recursive sorting
        result = mergeSortRecursive(result, reverse);
        
        return result;
    }
    
    /**
     * Recursive method for Merge Sort.
     * 
     * @param arr Array to be sorted
     * @param reverse Sort direction
     * @return Sorted array
     */
    private static int[] mergeSortRecursive(int[] arr, boolean reverse) {
        // Base case: if the array has size 0 or 1, it's already sorted
        if (arr.length <= 1) {
            return arr;
        }
        
        // Divide the array into two halves
        int mid = arr.length / 2;
        int[] left = Arrays.copyOfRange(arr, 0, mid);
        int[] right = Arrays.copyOfRange(arr, mid, arr.length);
        
        // Recursively sort each half
        left = mergeSortRecursive(left, reverse);
        right = mergeSortRecursive(right, reverse);
        
        // Merge the two sorted halves
        return merge(left, right, reverse);
    }
    
    /**
     * Method to merge two sorted sublists.
     * 
     * @param left Left sorted sublist
     * @param right Right sorted sublist
     * @param reverse Sort direction
     * @return Merged sorted array
     */
    private static int[] merge(int[] left, int[] right, boolean reverse) {
        int[] result = new int[left.length + right.length];
        int i = 0, j = 0, k = 0;
        
        // Compare elements from the two sublists and add the smaller (or larger) to the result
        while (i < left.length && j < right.length) {
            if (!reverse) {
                // Ascending order
                if (left[i] <= right[j]) {
                    result[k++] = left[i++];
                } else {
                    result[k++] = right[j++];
                }
            } else {
                // Descending order
                if (left[i] >= right[j]) {
                    result[k++] = left[i++];
                } else {
                    result[k++] = right[j++];
                }
            }
        }
        
        // Add the remaining elements from the left sublist
        while (i < left.length) {
            result[k++] = left[i++];
        }
        
        // Add the remaining elements from the right sublist
        while (j < right.length) {
            result[k++] = right[j++];
        }
        
        return result;
    }
    
    /**
     * Overloaded mergeSort method for ascending order by default.
     * 
     * @param arr Array to be sorted
     * @return Array sorted in ascending order
     */
    public static int[] mergeSort(int[] arr) {
        return mergeSort(arr, false);
    }
    
    /**
     * Implementation of the Merge Sort algorithm for strings.
     * 
     * @param arr String array to be sorted
     * @param reverse If true, sorts in descending order; if false, in ascending order
     * @return Sorted array
     */
    public static String[] mergeSort(String[] arr, boolean reverse) {
        // Create a copy of the array to avoid modifying the original
        String[] result = Arrays.copyOf(arr, arr.length);
        
        // Start the recursive sorting
        result = mergeSortRecursive(result, reverse);
        
        return result;
    }
    
    /**
     * Recursive method for Merge Sort of strings.
     * 
     * @param arr Array to be sorted
     * @param reverse Sort direction
     * @return Sorted array
     */
    private static String[] mergeSortRecursive(String[] arr, boolean reverse) {
        // Base case: if the array has size 0 or 1, it's already sorted
        if (arr.length <= 1) {
            return arr;
        }
        
        // Divide the array into two halves
        int mid = arr.length / 2;
        String[] left = Arrays.copyOfRange(arr, 0, mid);
        String[] right = Arrays.copyOfRange(arr, mid, arr.length);
        
        // Recursively sort each half
        left = mergeSortRecursive(left, reverse);
        right = mergeSortRecursive(right, reverse);
        
        // Merge the two sorted halves
        return merge(left, right, reverse);
    }
    
    /**
     * Method to merge two sorted string sublists.
     * 
     * @param left Left sorted sublist
     * @param right Right sorted sublist
     * @param reverse Sort direction
     * @return Merged sorted array
     */
    private static String[] merge(String[] left, String[] right, boolean reverse) {
        String[] result = new String[left.length + right.length];
        int i = 0, j = 0, k = 0;
        
        // Compare elements from the two sublists and add the smaller (or larger) to the result
        while (i < left.length && j < right.length) {
            if (!reverse) {
                // Ascending order
                if (left[i].compareTo(right[j]) <= 0) {
                    result[k++] = left[i++];
                } else {
                    result[k++] = right[j++];
                }
            } else {
                // Descending order
                if (left[i].compareTo(right[j]) >= 0) {
                    result[k++] = left[i++];
                } else {
                    result[k++] = right[j++];
                }
            }
        }
        
        // Add the remaining elements from the left sublist
        while (i < left.length) {
            result[k++] = left[i++];
        }
        
        // Add the remaining elements from the right sublist
        while (j < right.length) {
            result[k++] = right[j++];
        }
        
        return result;
    }
    
    /**
     * Overloaded mergeSort method for ascending order of strings by default.
     * 
     * @param arr String array to be sorted
     * @return Array sorted in ascending order
     */
    public static String[] mergeSort(String[] arr) {
        return mergeSort(arr, false);
    }
    
    /**
     * Main method to demonstrate the use of Merge Sort.
     */
    public static void main(String[] args) {
        // Example with numbers
        int[] arr = {64, 34, 25, 12, 22, 11, 90};
        System.out.println("Original array: " + Arrays.toString(arr));
        
        // Ascending order
        int[] sortedArr = mergeSort(arr);
        System.out.println("Ascending order: " + Arrays.toString(sortedArr));
        
        // Descending order
        int[] sortedArrDesc = mergeSort(arr, true);
        System.out.println("Descending order: " + Arrays.toString(sortedArrDesc));
        
        // Example with strings
        String[] strArr = {"banana", "apple", "orange", "pineapple", "grape"};
        System.out.println("\nOriginal string array: " + Arrays.toString(strArr));
        
        // Ascending order
        String[] sortedStrArr = mergeSort(strArr);
        System.out.println("Ascending order: " + Arrays.toString(sortedStrArr));
        
        // Descending order
        String[] sortedStrArrDesc = mergeSort(strArr, true);
        System.out.println("Descending order: " + Arrays.toString(sortedStrArrDesc));
    }
}

File: sorting/python/merge_sort.py
"""
Merge Sort - Sorting Algorithm

Time Complexity:
- Best case: O(n log n)
- Average case: O(n log n)
- Worst case: O(n log n)

Space Complexity: O(n) - requires additional space for temporary arrays

How it works:
Merge Sort is a divide and conquer sorting algorithm.
1. Divide the array into two halves
2. Recursively sort each half
3. Merge the two sorted halves to produce the final sorted array
"""

def merge_sort(arr, reverse=False):
    """
    Implementation of the Merge Sort algorithm.
    
    Args:
        arr: List to be sorted
        reverse: If True, sorts in descending order; if False, in ascending order
        
    Returns:
        Sorted list
    """
    # Create a copy of the array to avoid modifying the original
    result = arr.copy()
    
    # Internal function for recursive sorting
    def _merge_sort(arr):
        # Base case: if the array has size 0 or 1, it's already sorted
        if len(arr) <= 1:
            return arr
            
        # Divide the array into two halves
        mid = len(arr) // 2
        left = arr[:mid]
        right = arr[mid:]
        
        # Recursively sort each half
        left = _merge_sort(left)
        right = _merge_sort(right)
        
        # Merge the two sorted halves
        return _merge(left, right)
    
    # Function to merge two sorted sublists
    def _merge(left, right):
        result = []
        i = j = 0
        
        # Compare elements from the two sublists and add the smaller (or larger) to the result
        while i < len(left) and j < len(right):
            if not reverse:
                # Ascending order
                if left[i] <= right[j]:
                    result.append(left[i])
                    i += 1
                else:
                    result.append(right[j])
                    j += 1
            else:
                # Descending order
                if left[i] >= right[j]:
                    result.append(left[i])
                    i += 1
                else:
                    result.append(right[j])
                    j += 1
        
        # Add the remaining elements from the left sublist
        result.extend(left[i:])
        
        # Add the remaining elements from the right sublist
        result.extend(right[j:])
        
        return result
    
    # Start the recursive sorting
    return _merge_sort(result)

def main():
    """Main function to demonstrate the use of Merge Sort."""
    # Example with numbers
    arr = [64, 34, 25, 12, 22, 11, 90]
    print("Original array:", arr)
    
    # Ascending order
    sorted_arr = merge_sort(arr)
    print("Ascending order:", sorted_arr)
    
    # Descending order
    sorted_arr_desc = merge_sort(arr, reverse=True)
    print("Descending order:", sorted_arr_desc)
    
    # Example with strings
    str_arr = ["banana", "apple", "orange", "pineapple", "grape"]
    print("\nOriginal string array:", str_arr)
    
    # Ascending order
    sorted_str_arr = merge_sort(str_arr)
    print("Ascending order:", sorted_str_arr)
    
    # Descending order
    sorted_str_arr_desc = merge_sort(str_arr, reverse=True)
    print("Descending order:", sorted_str_arr_desc)

if __name__ == "__main__":
    main()
